### DISCUSSION TITLE
Build a visual editor that will mark rules, facts, predicates from a code base than configuring them

### DISCUSSION THREAD

**Discussion by ghost** [facebookincubator/Glean#386] [Ideas]
1. What's the theory behind this software - facts, predicates, rules. Why not build a visual builder where user will mark pieces of code with configuration than writing the configuring it. Let the code in form of the graph rerender with new configuration making user visually see what is needed and what is not needed.
2. This will also help to perform refactoring.

**Motivation** - I'm trying to build a web editor where I can perform frontend code refactoring from flow based into configuration based as the frontend code, when isolated away from the backend code is essentially a tree states of component's variants being shown or hidden.

**Comment by simonmar** [comment]
I'm afraid you'll need to elaborate this suggestion quite a bit, I can't grasp what you're getting at.

**Comment by ghost** [comment]
Currently in the glean project the configuration such as rule, fact, predictions are saved as a file in the backend and code analysis happens through service.

I like the project, and I'm building a code refactoring tool on the browser side without a backend. My design is based on few theories on my own such as code reduction, code to configuration, dead code removal and I'm still researching to lay the foundation.

Since glean is already doing it, I'm sure there are quire a lot of things I could learn from it to build something similar, from scratch in javascript or put community javascript projects pieces together to build my idea. So I was looking for information on theoritical concepts based on which glean is built up on so that I could study it, rewrite it in javascript.

To visualise my idea further, user will drop the code base on the browser with my UI that will parse the source, open tools for user to annotate symbols as tokens, delimiters, keywords and as next level, annotate tokens, delimiters, keywords as rules, and rules as facts. In this way, user don't have to type and store the rule/facts/ predictions to run against a code to validate it, the user can create them on fly. Plus it will give instant visual feedback.
